GX-TimeFrequency/src/minimax_grids.F90

```diff
@@ -0,0 +1,462 @@
+    allocate (cosft_tw(num_points, num_points))
+    allocate (sinft_wt(num_points, num_points))
+
+    ! get the weights for the cosine transform W^c(it) -> W^c(iw)
+    call get_transformation_weights(num_points, tau_points, omega_points, cosft_wt, e_min, e_max, &
+         max_errors(1), cos_t_to_cos_w, ierr)
+    if (ierr /= 0) return
+
+    ! get the weights for the cosine transform W^c(iw) -> W^c(it)
+    call get_transformation_weights(num_points, tau_points, omega_points, cosft_tw, e_min, e_max, &
+         max_errors(2), cos_w_to_cos_t, ierr)
+    if (ierr /= 0) return
+
+    ! get the weights for the sine transform Sigma^sin(it) -> Sigma^sin(iw) (PRB 94, 165109 (2016), Eq. 71)
+    call get_transformation_weights(num_points, tau_points, omega_points, sinft_wt, e_min, e_max, &
+         max_errors(3), sin_t_to_sin_w, ierr)
+    if (ierr /= 0) return
+
+    ! Compute the actual weights used for the inhomogeneous cosine/ FT and check whether
+    ! the two matrices for the forward/backward transform are the inverse of each other.
+    do i_point = 1, num_points
+       do j_point = 1, num_points
+          cosft_wt(j_point, i_point) = cosft_wt(j_point, i_point)*cos(tau_points(i_point)*omega_points(j_point))
+          cosft_tw(i_point, j_point) = cosft_tw(i_point, j_point)*cos(tau_points(i_point)*omega_points(j_point))
+          sinft_wt(j_point, i_point) = sinft_wt(j_point, i_point)*sin(tau_points(i_point)*omega_points(j_point))
+       end do
+    end do
+
+    allocate (mat(num_points, num_points))
+    mat(:, :) = matmul(cosft_wt, cosft_tw)
+    do i_point = 1, num_points
+       mat(i_point, i_point) = mat(i_point, i_point) - 1.0d0
+    end do
+    cosft_duality_error = maxval(abs(mat))
+
+    deallocate (mat)
+    deallocate (x_tw)
+
+  end subroutine gx_minimax_grid
+
+  ! **************************************************************************************************
+  ! /brief get the weights eiter for the cosine/sin transformation for tau to omega or viceversa
+  ! o num_points: Number of mesh points.
+  ! o tau_points: imaginary time grid points
+  ! o omega_points: imaginary frequency grid points
+  ! o weights: corresponding tranformation weights
+  ! o e_min: Minimum transition energy.
+  ! o e_max: Maximum transition energy.
+  ! o max_errors: Max error for the three kind of transforms
+  ! o transformation type : 1 the cosine transform cos(it) -> cos(iw)
+  !                       : 2 the cosine transform cos(iw) -> cos(it)
+  !                       : 3 the sine transform   sin(it) -> sin(iw)
+  ! o ierr: exit status
+  ! **************************************************************************************************
+  subroutine get_transformation_weights(num_points, tau_points, omega_points, weights, e_min, e_max, &
+       max_error, transformation_type, ierr)
+
+    integer, intent(in)                                :: num_points
+    real(kind=dp), allocatable, dimension(:), &
```

Comment:
I would be consistent, either real(dp) or real(kind=dp). It varies from routine to routine

Reply:
Done